### PR TITLE
Command: "add" command for all 3 types of task

### PR DESCRIPTION
**Pull request by nightism**
Now adding task commands of all 3 kinds of tasks can be done by one single keyword "add".
Parser for the new command is also updated.
- Parameters for adding event: "NAME" [STARTING_DATE] [STARTING_TIME] to [ENDING_DATE] [ENDING_TIME]
- Parameters for adding deadline: "NAME" [DATE] [TIME]
- Parameters for adding floating task: "NAME" [p-Priority]

Note that the "p-" in front of the Priority can not be omitted. The reason is indicated below in the example:

ADD a deadline: add "due name" 3
ADD a floating task: add "task name" p-3

If the p- is omitted, the two command will look the same.

### CODE REVIEW

**Comment by pyokagan** [comment]
@nightism Eh the build is not even passing though...

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+            return result;
+        }
+        if (isTime(args.get(0))) {
+            result.endTime = args.remove(0);
+        }
+
+        if (!args.isEmpty()) {
+            throw new IllegalValueException("wrong arguments given, please check the number and the order of the arguments");
+        }
+        return result;
```
Comment (pyokagan) [inline src/main/java/seedu/address/logic/parser/AddTaskParser.java:217]: Ugh, I thought that we were going to get a reusable parser that we can use for all commands (and implement new ones in the future), not coding everything manually...

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        try {
+            parser.parseTime(str);
+            return true;
+        } catch (IllegalValueException e) {
+            return false;
+        }
+    }
+
+    private static boolean isKeywordTo(String str) {
+        return str.trim().equals("to");
```
Comment (pyokagan) [inline src/main/java/seedu/address/logic/parser/AddTaskParser.java:258]: Is there a reason why this `trim()`, and thus this method, is needed?

Comment (nightism) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:258]: just ensure that the string passed in does not have extra white spaces

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java (hunk 1 above)
Comment (nightism) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:217]: ok, will think up a better idea of the pattern

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java (hunk 2 above)
Comment (pyokagan) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:258]: Doesn't the `ARG_PATTERN` already guarantee that?

Comment (pyokagan) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:258]: Unless the token is quoted, of course, but that is the user's problem.

**Comment by nightism** [comment]
@pyokagan sorry... I have no idea why the build is not passing...

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java (hunk 2 above)
Comment (nightism) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:258]: the `ARG_PATTERN` does not guarantee that

Comment (pyokagan) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:258]: Yes it does

``` java
package seedu.address.testutil;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.junit.Test;

public class TestTest {

    @Test
    public void test() {
        for (String x : splitArgs("   a   b   ")) {
            System.out.println("'" + x + "'");
        }
    }

    private static ArrayList<String> splitArgs(String str) {
        Pattern pat = Pattern.compile("\\s*\"(?<quotedArg>[^\"]+)\"\\s*|\\s*(?<unquotedArg>[^\\s]+)\\s*");
        final Matcher matcher = pat.matcher(str);
        final ArrayList<String> args = new ArrayList<>();
        int start = 0;
        while (matcher.find(start)) {
            args.add(matcher.group("quotedArg") != null ? matcher.group("quotedArg")
                                                        : matcher.group("unquotedArg"));
            start = matcher.end();
        }
        return args;
    }

}
```

```
'a'
'b'
```

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java (hunk 1 above)
Comment (pyokagan) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:217]: Can we find a solution that does not involve squashing all the parsing logic from `AddEventParser`, `AddDeadlineParser` etc. into one class?

Comment (pyokagan) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:217]: Something along the lines of this? (But there are lots of tiny details to iron out):

``` java
package seedu.address.logic.parser;

import seedu.address.logic.commands.Command;
import seedu.address.logic.commands.IncorrectCommand;

public class AddParser {

    private final AddFloatingTaskParser addFloatingTaskParser = new AddFloatingTaskParser();

    private final AddDeadlineParser addDeadlineParser = new AddDeadlineParser();

    private final AddEventParser addEventParser = new AddEventParser();

    public Command parse(String str) {
        Command cmd;

        cmd = addEventParser.parse(str);
        if (!(cmd instanceof IncorrectCommand)) {
            return cmd;
        }

        cmd = addDeadlineParser.parse(str);
        if (!(cmd instanceof IncorrectCommand)) {
            return cmd;
        }

        return addFloatingTaskParser.parse(str);
    }

}
```

File: src/main/java/seedu/address/logic/commands/AddTaskCommand.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            model.addDeadlineTask((DeadlineTask)task);
+        } else if (task instanceof FloatingTask) {
+            model.addFloatingTask((FloatingTask)task);
+        } else {
+            // should never reach this point
+        }
+        return new CommandResult(String.format(MESSAGE_SUCCESS, task));
+    }
+
+}
```
Comment (pyokagan) [inline src/main/java/seedu/address/logic/commands/AddTaskCommand.java:65]: What is the purpose of squashing all the `AddDeadlineCommand`, `AddEventCommand` etc. into one class?

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java (hunk 1 above)
Comment (nightism) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:217]: will try it out

File: src/main/java/seedu/address/logic/commands/AddTaskCommand.java (hunk 3 above)
Comment (pyokagan) [reply src/main/java/seedu/address/logic/commands/AddTaskCommand.java:65]: Why not just let the `AddTaskParser` return the correct `Command` type?

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java (hunk 1 above)
Comment (pyokagan) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:217]: Yep, let's go in that direction. The code needs to adhere to the Single Responsibility Principle, so each variation of the "add" command needs to have one class.

File: src/main/java/seedu/address/logic/commands/AddTaskCommand.java (hunk 3 above)
Comment (nightism) [reply src/main/java/seedu/address/logic/commands/AddTaskCommand.java:65]: Yes, just adhere to the Single Responsibility Principle, and just let the "add" command parser return different command types

File: src/main/java/seedu/address/logic/parser/AddTaskParser.java (hunk 2 above)
Comment (nightism) [reply src/main/java/seedu/address/logic/parser/AddTaskParser.java:258]: OK, I see... will improve this part

**Comment by nightism** [comment]
closed by creating a new PR

**Pull request closed without merging**